examples/auction/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+    BidTooLow { bidded: Amount, highest_bid: Amount }, // raised if bid is lower than highest amount
+    BidsOverWaitingForAuctionFinalization, // raised if bid is placed after auction expiry time
+    AuctionFinalized, // raised if bid is placed after auction has been finalized
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum FinalizeError {
+    BidMapError, // raised if there is a mistake in the bid map that keeps track of all accounts' bids
+    AuctionStillActive, // raised if there is an attempt to finalize the auction before its expiry
+    AuctionFinalized, // raised if there is an attempt to finalize an already finalized auction
+}
+
+#[init(contract = "auction", parameter = "InitParameter")]
+fn auction_init(ctx: &impl HasInitContext) -> InitResult<State> {
+    let parameter: InitParameter = ctx.parameter_cursor().get()?;
+    Ok(fresh_state(parameter.item, parameter.expiry))
+}
+
+#[receive(contract = "auction", name = "bid", payable)]
+fn auction_bid<A: HasActions>(
+    ctx: &impl HasReceiveContext,
+    amount: Amount,
+    state: &mut State,
+) -> Result<A, ReceiveError> {
+    ensure!(state.auction_state <= AuctionState::BidsOverWaitingForAuctionFinalization, ReceiveError::AuctionFinalized);
+
+    let slot_time = ctx.metadata().slot_time();
+    if slot_time > state.expiry {
+        state.auction_state = AuctionState::BidsOverWaitingForAuctionFinalization;
+        bail!(ReceiveError::BidsOverWaitingForAuctionFinalization);
+    }
+
+    let sender_address = match ctx.sender() {
+        Address::Contract(_) => bail!(ReceiveError::ContractSender),
+        Address::Account(account_address) => account_address,
+    };
+    let existing_bid = *state.bids.entry(sender_address).or_insert_with(|| Amount::zero());
+
+    let new_bid = existing_bid + amount;
+    // Ensure that the new bid exceeds the highest bid so far
+    ensure!(new_bid > state.highest_bid, ReceiveError::BidTooLow { bidded: amount, highest_bid: state.highest_bid });
+    state.highest_bid = new_bid;
+    state.bids.insert(sender_address, existing_bid + amount);
+
+    Ok(A::accept())
+}
+
+#[receive(contract = "auction", name = "finalize")]
+fn auction_finalize<A: HasActions>(
+    ctx: &impl HasReceiveContext,
+    state: &mut State,
+) -> Result<A, FinalizeError> {
+    ensure!(state.auction_state <= AuctionState::BidsOverWaitingForAuctionFinalization, FinalizeError::AuctionFinalized);
+
+    let slot_time = ctx.metadata().slot_time();
+    if slot_time <= state.expiry {
+        bail!(FinalizeError::AuctionStillActive);
+    }
+
+    let owner = ctx.owner();
+
+    ensure!(state.auction_state == AuctionState::Active, FinalizeError::AuctionFinalized);
+
+    let balance = ctx.self_balance();
+    if balance == Amount::zero() {
+        Ok(A::accept())
+    } else {
+        let mut return_action = A::simple_transfer(&owner, state.highest_bid);
+        let mut remaining_bids = BTreeMap::new();
```

Comment:
I would not use a map here since it can have exactly one element.

I think the better thing to do would be to initialize with `None` and then ...